File: app/src/main/java/com/proxy/shadowsocksr/impl/SSRTunnel.java
package com.proxy.shadowsocksr.impl;

import android.util.Log;

import com.proxy.shadowsocksr.impl.interfaces.OnNeedProtectTCPListener;
import com.proxy.shadowsocksr.impl.plugin.obfs.AbsObfs;
import com.proxy.shadowsocksr.impl.plugin.obfs.ObfsChooser;
import com.proxy.shadowsocksr.impl.plugin.proto.AbsProtocol;
import com.proxy.shadowsocksr.impl.plugin.proto.ProtocolChooser;

import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.UnknownHostException;
import java.nio.ByteBuffer;
import java.nio.channels.ServerSocketChannel;
import java.nio.channels.SocketChannel;
import java.util.HashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public final class SSRTunnel extends Thread
{
    private ServerSocketChannel ssc;
    private final String remoteIP;
    private final String localIP;
    private final int remotePort;
    private final int localPort;
    private final int dnsPort;
    private final String pwd;
    private final String cryptMethod;
    private final String tcpProtocol;
    private final String obfsMethod;
    private final String obfsParam;

    private byte[] dnsIp;

    private ExecutorService localThreadPool;
    private ExecutorService remoteThreadPool;

    private volatile boolean isRunning = true;

    private OnNeedProtectTCPListener onNeedProtectTCPListener;

    private final HashMap<String, Object> shareParam;

    public SSRTunnel(String remoteIP, String localIP, String dnsIP, int remotePort,
            int localPort, int dnsPort, String cryptMethod, String tcpProtocol,
            String obfsMethod, String obfsParam, String pwd)
    {
        this.remoteIP = remoteIP;
        this.localIP = localIP;
        this.remotePort = remotePort;
        this.localPort = localPort;
        this.dnsPort = dnsPort;
        this.cryptMethod = cryptMethod;
        this.pwd = pwd;
        this.tcpProtocol = tcpProtocol;
        this.obfsMethod = obfsMethod;
        this.obfsParam = obfsParam;

        shareParam = new HashMap<>();

        try
        {
            dnsIp = InetAddress.getByName(dnsIP).getAddress();
        }
        catch (UnknownHostException ignored)
        {
        }
    }

    public void setOnNeedProtectTCPListener(
            OnNeedProtectTCPListener onNeedProtectTCPListener)
    {
        this.onNeedProtectTCPListener = onNeedProtectTCPListener;
    }

    class ChannelAttach
    {
        public ByteBuffer localReadBuf = ByteBuffer.allocate(8192);
        public ByteBuffer remoteReadBuf = ByteBuffer.allocate(8192);
        public TCPEncryptor crypto = new TCPEncryptor(pwd, cryptMethod);
        public AbsObfs obfs = ObfsChooser
                .getObfs(obfsMethod, remoteIP, remotePort, 1440, obfsParam);
        public AbsProtocol proto = ProtocolChooser
                .getProtocol(tcpProtocol, remoteIP, remotePort, 1440, shareParam);
        public SocketChannel localSkt;
        public SocketChannel remoteSkt;
    }

    @Override public void run()
    {
        localThreadPool = Executors.newCachedThreadPool();
        remoteThreadPool = Executors.newCachedThreadPool();
        //new ThreadPoolExecutor(1, Integer.MAX_VALUE, 300L, TimeUnit.SECONDS, new SynchronousQueue<Runnable>());

        while (isRunning)//When tcp server crashed, restart it.
        {
            try
            {
                ssc = ServerSocketChannel.open();
                //default is block
                ssc.socket().bind(new InetSocketAddress(localIP, localPort));
                while (isRunning)
                {
                    ChannelAttach attach = new ChannelAttach();
                    attach.localSkt = ssc.accept();
                    localThreadPool.execute(new LocalSocketHandler(attach));
                }
            }
            catch (Exception ignored)
            {
            }
            try
            {
                ssc.close();
            }
            catch (Exception ignored)
            {
            }
        }
    }

    class LocalSocketHandler implements Runnable
    {
        private final ChannelAttach attach;

        public LocalSocketHandler(ChannelAttach attach)
        {
            this.attach = attach;
        }

        @Override public void run()
        {
            try
            {
                //default is block
                attach.localSkt.socket().setTcpNoDelay(true);
                attach.localSkt.socket().setReuseAddress(true);
                //
                if (!prepareRemote(attach, remoteIP, remotePort))
                {
                    Log.e("EXC", "REMOTE CONNECT FAILED!");
                    return;
                }
                //
                attach.localReadBuf.put((byte) 1).put(dnsIp).put((byte) ((dnsPort >> 8) & 0xFF))
                                   .put((byte) (dnsPort & 0xFF));
                //
                remoteThreadPool.execute(new RemoteSocketHandler(attach));
                //
                while (isRunning)
                {
                    if (!checkSessionAlive(attach))
                    {
                        Log.e("EXC", "DEAD");
                        break;
                    }
                    int rcnt = attach.localSkt.read(attach.localReadBuf);
                    if (rcnt < 1)
                    {
                        break;
                    }

                    byte[] recv = new byte[attach.localReadBuf.flip().limit()];
                    attach.localReadBuf.get(recv);
                    //
                    recv = attach.proto.beforeEncrypt(recv);
                    recv = attach.crypto.encrypt(recv);
                    recv = attach.obfs.afterEncrypt(recv);
                    //
                    attach.remoteSkt.write(ByteBuffer.wrap(recv));
                    attach.localReadBuf.clear();
                }
            }
            catch (Exception ignored)
            {
            }
            cleanSession(attach);
        }
    }

    private boolean prepareRemote(ChannelAttach attach, String remoteIP, int remotePort)
    throws Exception
    {
        attach.remoteSkt = SocketChannel.open();
        //default is block
        attach.remoteSkt.socket().setReuseAddress(true);
        attach.remoteSkt.socket().setTcpNoDelay(true);
        boolean success = onNeedProtectTCPListener.onNeedProtectTCP(attach.remoteSkt.socket());
        if (!success)
        {
            return false;
        }
        attach.remoteSkt.connect(new InetSocketAddress(remoteIP, remotePort));
        return attach.remoteSkt.isConnected();
    }

    private boolean checkSessionAlive(ChannelAttach attach)
    {
        return attach.localSkt != null &&
               attach.remoteSkt != null;
    }

    private void cleanSession(ChannelAttach attach)
    {
        try
        {
            attach.remoteSkt.close();
            attach.localSkt.close();
        }
        catch (Exception ignored)
        {
        }
        attach.remoteSkt = null;
        attach.localSkt = null;
        attach.obfs = null;
        attach.proto = null;
        attach.crypto = null;
        attach.localReadBuf = null;
        attach.remoteReadBuf = null;
    }

    public void stopTunnel()
    {
        isRunning = false;
        try
        {
            ssc.close();
        }
        catch (Exception ignored)
        {
        }
        localThreadPool.shutdown();
        ssc = null;
    }

    class RemoteSocketHandler implements Runnable
    {
        private final ChannelAttach attach;

        public RemoteSocketHandler(ChannelAttach attach)
        {
            this.attach = attach;
        }

        @Override public void run()
        {
            try
            {
                while (isRunning)
                {
                    if (!checkSessionAlive(attach))
                    {
                        Log.e("EXC", "DEAD");
                        break;
                    }
                    int rcnt = attach.remoteSkt.read(attach.remoteReadBuf);
                    if (rcnt < 0)
                    {
                        break;
                    }
                    attach.remoteReadBuf.flip();
                    byte[] recv = new byte[rcnt];
                    attach.remoteReadBuf.get(recv);
                    //
                    recv = attach.obfs.beforeDecrypt(recv, false);//TODO
                    recv = attach.crypto.decrypt(recv);
                    recv = attach.proto.afterDecrypt(recv);
                    //
                    attach.localSkt.write(ByteBuffer.wrap(recv));
                    attach.remoteReadBuf.clear();
                }
            }
            catch (Exception ignored)
            {
            }
            cleanSession(attach);
        }
    }
}

File: app/src/main/java/com/proxy/shadowsocksr/impl/SSRTunnel.kt
package com.proxy.shadowsocksr.impl

import android.util.Log

import com.proxy.shadowsocksr.impl.interfaces.OnNeedProtectTCPListener
import com.proxy.shadowsocksr.impl.plugin.obfs.AbsObfs
import com.proxy.shadowsocksr.impl.plugin.obfs.ObfsChooser
import com.proxy.shadowsocksr.impl.plugin.proto.AbsProtocol
import com.proxy.shadowsocksr.impl.plugin.proto.ProtocolChooser

import java.net.InetAddress
import java.net.InetSocketAddress
import java.net.UnknownHostException
import java.nio.ByteBuffer
import java.nio.channels.ServerSocketChannel
import java.nio.channels.SocketChannel
import java.util.HashMap
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

class SSRTunnel(private val remoteIP: String, private val localIP: String, dnsIP: String, private val remotePort: Int,
                private val localPort: Int, private val dnsPort: Int, private val cryptMethod: String, private val tcpProtocol: String,
                private val obfsMethod: String, private val obfsParam: String, private val pwd: String, private var isVPN: Boolean) : Thread()
{
    private var ssc: ServerSocketChannel? = null

    private var dnsIp: ByteArray? = null

    private var localThreadPool: ExecutorService? = null
    private var remoteThreadPool: ExecutorService? = null

    @Volatile private var isRunning = true

    private var onNeedProtectTCPListener: OnNeedProtectTCPListener? = null

    private val shareParam: HashMap<String, Any> = hashMapOf()

    init
    {
        try
        {
            dnsIp = InetAddress.getByName(dnsIP).address
        }
        catch (ignored: UnknownHostException)
        {
        }
    }

    fun setOnNeedProtectTCPListener(
            onNeedProtectTCPListener: OnNeedProtectTCPListener)
    {
        this.onNeedProtectTCPListener = onNeedProtectTCPListener
    }

    internal inner class ChannelAttach
    {
        var localReadBuf: ByteBuffer? = ByteBuffer.allocate(8192)
        var remoteReadBuf: ByteBuffer? = ByteBuffer.allocate(8192)
        var crypto: TCPEncryptor? = TCPEncryptor(pwd, cryptMethod)
        var obfs: AbsObfs? = ObfsChooser.getObfs(obfsMethod, remoteIP, remotePort, 1440, obfsParam)
        var proto: AbsProtocol? = ProtocolChooser.getProtocol(tcpProtocol, remoteIP, remotePort,
                1440, shareParam)
        var localSkt: SocketChannel? = null
        var remoteSkt: SocketChannel? = null
    }

    override fun run()
    {
        localThreadPool = Executors.newCachedThreadPool()
        remoteThreadPool = Executors.newCachedThreadPool()
        //new ThreadPoolExecutor(1, Integer.MAX_VALUE, 300L, TimeUnit.SECONDS, new SynchronousQueue<Runnable>());

        while (isRunning)
        //When tcp server crashed, restart it.
        {
            try
            {
                ssc = ServerSocketChannel.open()
                //default is block
                ssc!!.socket().bind(InetSocketAddress(localIP, localPort))
                while (isRunning)
                {
                    val attach = ChannelAttach()
                    attach.localSkt = ssc!!.accept()
                    localThreadPool!!.execute(LocalSocketHandler(attach))
                }
            }
            catch (ignored: Exception)
            {
            }

            try
            {
                ssc!!.close()
            }
            catch (ignored: Exception)
            {
            }

        }
    }

    internal inner class LocalSocketHandler(private val attach: ChannelAttach) : Runnable
    {
        override fun run()
        {
            try
            {
                //default is block
                attach.localSkt!!.socket().tcpNoDelay = true
                attach.localSkt!!.socket().reuseAddress = true
                //
                if (!prepareRemote(attach, remoteIP, remotePort))
                {
                    Log.e("EXC", "REMOTE CONNECT FAILED!")
                    return
                }
                //
                attach.localReadBuf!!.put((1).toByte()).put(dnsIp).put(
                        ((dnsPort shr 8) and 0xFF).toByte()).put((dnsPort and 0xFF).toByte())
                //
                remoteThreadPool!!.execute(RemoteSocketHandler(attach))
                //
                while (isRunning)
                {
                    if (!checkSessionAlive(attach))
                    {
                        Log.e("EXC", "DEAD")
                        break
                    }
                    val rcnt = attach.localSkt!!.read(attach.localReadBuf)
                    if (rcnt < 1)
                    {
                        break
                    }

                    var recv = ByteArray(attach.localReadBuf!!.flip().limit())
                    attach.localReadBuf!!.get(recv)
                    //
                    recv = attach.proto!!.beforeEncrypt(recv)
                    recv = attach.crypto!!.encrypt(recv)
                    recv = attach.obfs!!.afterEncrypt(recv)
                    //
                    attach.remoteSkt!!.write(ByteBuffer.wrap(recv))
                    attach.localReadBuf!!.clear()
                }
            }
            catch (ignored: Exception)
            {
            }

            cleanSession(attach)
        }
    }

    @Throws(Exception::class)
    private fun prepareRemote(attach: ChannelAttach, remoteIP: String, remotePort: Int): Boolean
    {
        attach.remoteSkt = SocketChannel.open()
        //default is block
        attach.remoteSkt!!.socket().reuseAddress = true
        attach.remoteSkt!!.socket().tcpNoDelay = true
        if(isVPN)
        {
            val success = onNeedProtectTCPListener!!.onNeedProtectTCP(attach.remoteSkt!!.socket())
            if (!success)
            {
                return false
            }
        }
        attach.remoteSkt!!.connect(InetSocketAddress(remoteIP, remotePort))
        return attach.remoteSkt!!.isConnected
    }

    private fun checkSessionAlive(attach: ChannelAttach): Boolean
    {
        return attach.localSkt != null && attach.remoteSkt != null
    }

    private fun cleanSession(attach: ChannelAttach)
    {
        try
        {
            attach.remoteSkt!!.close()
            attach.localSkt!!.close()
        }
        catch (ignored: Exception)
        {
        }

        attach.remoteSkt = null
        attach.localSkt = null
        attach.obfs = null
        attach.proto = null
        attach.crypto = null
        attach.localReadBuf = null
        attach.remoteReadBuf = null
    }

    fun stopTunnel()
    {
        isRunning = false
        try
        {
            ssc!!.close()
        }
        catch (ignored: Exception)
        {
        }

        localThreadPool!!.shutdown()
        ssc = null
    }

    internal inner class RemoteSocketHandler(private val attach: ChannelAttach) : Runnable
    {

        override fun run()
        {
            try
            {
                while (isRunning)
                {
                    if (!checkSessionAlive(attach))
                    {
                        Log.e("EXC", "DEAD")
                        break
                    }
                    val rcnt = attach.remoteSkt!!.read(attach.remoteReadBuf)
                    if (rcnt < 0)
                    {
                        break
                    }
                    attach.remoteReadBuf!!.flip()
                    var recv = ByteArray(rcnt)
                    attach.remoteReadBuf!!.get(recv)
                    //
                    recv = attach.obfs!!.beforeDecrypt(recv, false)//TODO
                    recv = attach.crypto!!.decrypt(recv)
                    recv = attach.proto!!.afterDecrypt(recv)
                    //
                    attach.localSkt!!.write(ByteBuffer.wrap(recv))
                    attach.remoteReadBuf!!.clear()
                }
            }
            catch (ignored: Exception)
            {
            }

            cleanSession(attach)
        }
    }
}
